### PR TITLE
Add new formatted answer column to ORA2

### PR DESCRIPTION
**Pull request by vagupta16**
The ORA2 responses collection process currently does not correctly decode the raw answers retrieved from MySQL, which are JSON strings.

ORA2 workflow:

lms/djangoapps/instructor_task/api.py:submit_ora2_request_task() -->
lms/djangoapps/instructor_task/tasks.py:get_ora2_responses() -->
lms/djangoapps/instructor_task/tasks_helper.py:push_ora2_responses_to_s3() -->

the above calls the shared method API at lms/djangoapps/instructor_task/tasks_helper.py:_push_csv_responses_to_s3() using the csv_fn titled collect_email_ora2_data or collect_anonymous_ora2_data, both of which are located at lms/djangoapps/instructor/utils.py

The above methods then call lms/djangoapps/instructor/collect_ora2_data(), which returns the relevant header and rows.

This header and rows are pushed by _push_csv_responses_to_s3 to a ReportStore (lms/djangoapps/instructor_task/models.py).

To avoid modifying shared code, I modified the collect_ora2_data() method to actually iterate through the header and all of the rows to append a formatted, true Unicode version of the raw answer.

NOTE: Still looking for a good way to actually test this.

### CODE REVIEW

File: lms/djangoapps/instructor/utils.py [hunk 1]
```diff
@@ -186,7 +187,13 @@ def collect_ora2_data(course_id, include_email=False):
     cursor.execute(raw_queries[0])
     cursor.execute(raw_queries[1], [course_id])
     header = [item[0] for item in cursor.description]
-    return header, cursor.fetchall()
+    header.append('formatted_answer')
+    raw_answer_index = 4
```
Comment (vagupta16) [inline lms/djangoapps/instructor/utils.py]: I'm really not a fan of this, but I'm not sure how else to do this. The query upon which this function relies returns the raw_answer as the fifth column (thus index 4).

File: lms/djangoapps/instructor/utils.py [hunk 2]
```diff
@@ -186,7 +187,13 @@ def collect_ora2_data(course_id, include_email=False):
...
+    data = cursor.fetchall()
+    data_rows = [list(row) + [dumps(
+        loads(row[raw_answer_index]),
+        ensure_ascii=False)] for row in data]
```
Comment (vagupta16) [inline lms/djangoapps/instructor/utils.py]: This line does the following:

for each row, grab the raw_answer (which is a JSON string), load it as a JSON object (a Python map), and then re-serialize it as a JSON object (without enforcing ASCII constraints to ensure that the Unicode characters are not escaped)

**Comment by vagupta16** [comment]
test this please

File: lms/djangoapps/instructor/utils.py [hunk 3]
```diff
@@ -186,7 +187,13 @@ def collect_ora2_data(course_id, include_email=False):
     cursor.execute(raw_queries[0])
     cursor.execute(raw_queries[1], [course_id])
     header = [item[0] for item in cursor.description]
-    return header, cursor.fetchall()
+    header.append('formatted_answer')
+    raw_answer_index = header.index('raw_answer')
+    data = cursor.fetchall()
+    data_rows = [list(row) + [dumps(
```
Comment (stvstnfrd) [inline lms/djangoapps/instructor/utils.py]: Instead, format like:

```python
data_rows = [
    list(row) + [
        dumps(
            loads(row[raw_answer_index]),
            ensure_ascii=False,
        )
    ]
    for row in data
]
```

**Comment by stvstnfrd** [comment]
@dcadams Would you mind putting eyes on this too?

File: lms/djangoapps/instructor/utils.py (hunk 3 above)
Comment (caesar2164) [inline lms/djangoapps/instructor/utils.py]: yes please, that's much easier to understand!

Comment (stvstnfrd) [inline lms/djangoapps/instructor/utils.py]: And just as a general note, Python comprehensions are easiest to follow with the format:

```python
comprehension = [
    item
    for item in collection
    if condition
]
# where condition is an optional boolean expression
```

**Comment by vagupta16** [comment]
@stvstnfrd Here's what the CSV will look like: 

https://docs.google.com/a/stanford.edu/spreadsheets/d/19lVUsSfghs1-K4M9ZsUfdKlkHiUWXtp1lYwHQx5o9Hg/edit?usp=sharing

**Comment by dcadams** [comment]
It's a bit weird that Google docs displays the whole field backwards for me but the csv file looks correct and it displays ok in LibreCalc. I see this doesn't have test coverage but neither does the rest of this report and I think it would be a bit onerous to manually insert into all those tables so I'm :+1:

**Comment by vagupta16** [comment]
@dcadams Pretty sure that Drive recognizes the Arabic and then displays it as right-to-left (which I guess is the correct behavior?).

**Comment by vagupta16** [comment]
test this please

**Comment by vagupta16** [comment]
test this please

**Comment by vagupta16** [comment]
test this please

**Comment by vagupta16** [comment]
@stvstnfrd

**Comment by stvstnfrd** [comment]
@vagupta16 As mentioned F2F...

This technically fulfills the request from the story. That said, I'm wondering if we can/should better format this output further..

We're still dumping a JSON blob which isn't particularly clear to instructors (they often complain about JSON and want to just see the answer).

Could we extract just the plaintext part, something like:

```python
# Mock some test data
# TODO: Confirm if my assumptions about data format are correct
raw_answer_index = 1
data = [
    [
        'fluff',
        {
            'parts': [
                {
                    'text': 'raw_answer-%s-%s' % (i, j),
                }
                for j in xrange(3)
            ],
        },
    ]
    for i in xrange(3)
]

# Define a helper to extract the formatted data from a row
def extract_answer(row):
    raw_answer = row[raw_answer_index]
    parts = raw_answer.get('parts', [])
    answer = [
        part.get('text', '')
        for part in parts
    ]
    answer = ' '.join(answer)
    return answer

# Append the formatted answer to the row
for row in data:
    row.append(extract_answer(row))

# Debug to verify
print(data)
```

**Comment by vagupta16** [comment]
@stvstnfrd Comment above makes sense and should be an easy fix. One thing I still am unclear about is what multiple "parts" actually represent. I've created mock peer assessment questions locally and answered them, but I have not yet been able to reproduce test answers that have multiple parts. 

I bring this up mostly from a standpoint of how we join answers from multiple parts. Perhaps rather than joining with a space, we could join with a newline character?

**Comment by dcadams** [comment]
I may be able to show you what multiple parts are.

**Comment by vagupta16** [comment]
@stvstnfrd Looks like each part corresponds to an answer to a different prompt. I was thinking about perhaps adding a column a part, but within at least the CDDRL csv, different raw answers actually have different numbers of parts. That seems to make adding new columns problematic. 

I'm not sure how to cleanly solve this without just concatenating all the parts together with newlines in between.

**Comment by stvstnfrd** [comment]
@vagupta16 I originally considered that.

The issue there is that this is a CSV, so I'm not sure displaying newlines really helps us; how would that render? I don't think we want to display `\n` literally in a CSV and CSVs don't lend to multiline display.

I'm not sure there is a perfect solution here. What if a response itself contains a newline?
So it seems like we have some cases that may never be 100% explicit about the data format, but I think that's ok.

Given that we still provide the raw JSON, I'm inclined to say this is fine because the instructor still has access to the raw data anyway.

The real thing instructors want here (and often complain about), is that it's difficult to see the answer. I'd be open to some other separator, ` -- ` or some such, but I don't think newline is a good fit for a CSV.

Thoughts?

**Comment by vagupta16** [comment]
Modified code to make formatted answer not be JSON. 

Currently uses '\n' as a separator. Looks to appear fine in Google Sheets - compare formatted answer here with '\n' separator (https://docs.google.com/a/stanford.edu/spreadsheets/d/1up7t5XWPk2PyUjuF72zZDLeGG-YpoOIj47Poc7Dp73I/edit?usp=sharing) to formatted answer without '\n' separator (https://docs.google.com/a/stanford.edu/spreadsheets/d/1Prv0QmbVyfDPq1hY73mlpgiQxO06Hijd4MOxbbbf0SI/edit?usp=sharing). 

I don't know how it looks in Excel. The raw '\n' that you see in the answer are leftovers from the raw answer - currently not sure how to take care of those.

**Comment by varunabgupta** [comment]
@stvstnfrd

**Comment by vagupta16** [comment]
@stvstnfrd Experimenting with other separators as per your earlier suggestion

File: lms/djangoapps/instructor/utils.py [hunk 4]
```diff
@@ -186,7 +187,26 @@ def collect_ora2_data(course_id, include_email=False):
...
+    return header, data_rows
+
+
+def extract_answer(raw_answer):
+    parts = raw_answer.get('parts', [])
+    answer = [
+        part.get('text', '')
+        for part in parts
+    ]
+    answer = '\n'.join(answer)
```
Comment (vagupta16) [inline lms/djangoapps/instructor/utils.py]: Place the chosen separator here - '<separator>'.join(answer)

Comment (stvstnfrd) [inline lms/djangoapps/instructor/utils.py]: Yeah, we should probably just stick this in a constant for ease of updating.

```python
ORA2_ANSWER_PART_SEPARATOR = '\n-----\n'
# ...
answer = ORA2_ANSWER_PART_SEPARATOR.join(answer)
```

**Comment by stvstnfrd** [comment]
@vagupta16 I like that final option best; it makes it obvious that they didn't just hit `ENTER` during their response, but that these really are distinct parts.

**Comment by vagupta16** [comment]
@stvstnfrd Pushed, here's what '\n-----\n' looks like:

https://docs.google.com/a/stanford.edu/spreadsheets/d/1BMNHgCvR9DvS7J3F-g6A-X0SZXmAAMqmyLerLybpXq4/edit?usp=sharing

**Comment by stvstnfrd** [comment]
Cool, +1 pending tests passing.

**Comment by stvstnfrd** [comment]
retest this please

**Pull request merged**